File: packages/render/core/src/animation.ts
import type {
  ActiveAnimationProjection,
  AnimationTime,
  ResolvedAnimationTrackProjection,
  ViewBackgroundLayerProjection,
  ViewBackgroundProjection,
  ViewCharacterProjection,
} from './index'

export interface ProjectedTrackValue {
  property: string
  value: unknown
}

export function collectTrackValues(
  animations: readonly Readonly<ActiveAnimationProjection>[],
  target: string,
  now: number,
): ProjectedTrackValue[] {
  const collected: ProjectedTrackValue[] = []
  for (const animation of animations) {
    const elapsed = resolveAnimationLocalTime(animation, now)
    if (elapsed === undefined)
      continue
    for (const track of animation.resolvedTracks) {
      if (track.target !== target)
        continue
      const value = resolveTrackValue(track, elapsed, animation.duration)
      if (value !== undefined) {
        collected.push({ property: track.property, value })
      }
    }
  }
  return collected
}

export function applyTrackValues(target: Record<string, unknown>, tracks: readonly ProjectedTrackValue[]): void {
  for (const track of tracks) {
    setPath(target, track.property, track.value)
  }
}

export function cloneCharacter(character: Readonly<ViewCharacterProjection>): ViewCharacterProjection {
  return {
    ...character,
    position: character.position ? { ...character.position } : undefined,
    metadata: character.metadata ? { ...character.metadata } : undefined,
  }
}

export function cloneBackground(background: Readonly<ViewBackgroundProjection>): ViewBackgroundProjection {
  return {
    ...background,
    transition: background.transition ? { ...background.transition } : undefined,
    video: background.video
      ? {
          ...background.video,
          transition: background.video.transition ? { ...background.video.transition } : undefined,
          metadata: background.video.metadata ? { ...background.video.metadata } : undefined,
        }
      : undefined,
    layers: background.layers?.map(layer => cloneBackgroundLayer(layer)),
    composition: background.composition ? cloneUnknownRecord(background.composition) : undefined,
    metadata: background.metadata ? { ...background.metadata } : undefined,
  }
}

export function cloneBackgroundLayer(layer: Readonly<ViewBackgroundLayerProjection>): ViewBackgroundLayerProjection {
  return {
    ...layer,
    composition: layer.composition ? cloneUnknownRecord(layer.composition) : undefined,
    transition: layer.transition ? { ...layer.transition } : undefined,
    metadata: layer.metadata ? { ...layer.metadata } : undefined,
  }
}

function resolveAnimationLocalTime(animation: Readonly<ActiveAnimationProjection>, now: number): number | undefined {
  if (animation.state === 'stopped' && animation.endedAt === undefined) {
    return undefined
  }
  const playbackRate = Number.isFinite(animation.playbackRate) && animation.playbackRate > 0 ? animation.playbackRate : 1
  const delay = Math.max(0, animation.delay ?? 0)
  const duration = Math.max(0, animation.duration)
  const fill = animation.fill ?? 'forwards'
  const wallElapsed = (animation.state === 'paused' && animation.pausedAt !== undefined)
    ? animation.pausedAt - animation.startedAt
    : (animation.state === 'stopped' && animation.endedAt !== undefined)
        ? animation.endedAt - animation.startedAt
        : now - animation.startedAt
  const timelineElapsed = Math.max(0, wallElapsed * playbackRate)
  if (timelineElapsed < delay) {
    return fill === 'backwards' || fill === 'both'
      ? applyAnimationDirection(0, duration, 0, animation.direction)
      : undefined
  }

  if (duration === 0) {
    return 0
  }

  const activeElapsed = timelineElapsed - delay
  const loopCount = animation.loop === true
    ? Infinity
    : typeof animation.loop === 'number'
      ? Math.max(1, animation.loop)
      : 1
  const totalDuration = duration * loopCount

  if (activeElapsed >= totalDuration) {
    if (animation.loop === true) {
      const local = activeElapsed % duration
      const iteration = Math.floor(activeElapsed / duration)
      return applyAnimationDirection(local, duration, iteration, animation.direction)
    }

    if (fill === 'forwards' || fill === 'both') {
      return applyAnimationDirection(duration, duration, loopCount - 1, animation.direction)
    }
    return undefined
  }

  const iteration = Math.floor(activeElapsed / duration)
  const local = activeElapsed - iteration * duration
  return applyAnimationDirection(local, duration, iteration, animation.direction)
}

function applyAnimationDirection(
  local: number,
  duration: number,
  iteration: number,
  direction: Readonly<ActiveAnimationProjection>['direction'],
): number {
  const normalized = direction ?? 'normal'
  const reversed = normalized === 'reverse'
    || (normalized === 'alternate' && iteration % 2 === 1)
    || (normalized === 'alternate-reverse' && iteration % 2 === 0)
  return reversed ? duration - local : local
}

function resolveTrackValue(
  track: Readonly<ResolvedAnimationTrackProjection>,
  elapsed: number,
  duration: number,
): unknown {
  const keyframes = [...track.keyframes].sort((left, right) => resolveAt(left.at, duration) - resolveAt(right.at, duration))
  if (keyframes.length === 0)
    return undefined
  if (keyframes.length === 1)
    return keyframes[0].value

  const first = keyframes[0]
  const firstAt = resolveAt(first.at, duration)
  if (elapsed <= firstAt)
    return first.value

  for (let index = 1; index < keyframes.length; index++) {
    const previous = keyframes[index - 1]
    const next = keyframes[index]
    const previousAt = resolveAt(previous.at, duration)
    const nextAt = resolveAt(next.at, duration)
    if (elapsed <= nextAt) {
      if (elapsed === nextAt)
        return next.value
      if (track.interpolation === 'step' || track.interpolation === 'discrete')
        return previous.value
      const progress = nextAt === previousAt ? 1 : (elapsed - previousAt) / (nextAt - previousAt)
      return interpolateTrackValue(previous.value, next.value, easeProgress(progress, next.easing ?? previous.easing), track.interpolation)
    }
  }

  return keyframes[keyframes.length - 1].value
}

function interpolateTrackValue(
  from: unknown,
  to: unknown,
  progress: number,
  interpolation: Readonly<ResolvedAnimationTrackProjection>['interpolation'],
): unknown {
  if (typeof from === 'number' && typeof to === 'number') {
    return from + (to - from) * progress
  }

  if ((interpolation === 'array' || interpolation === 'vector') && Array.isArray(from) && Array.isArray(to)) {
    return interpolateArray(from, to, progress)
  }

  if ((interpolation === 'vector' || interpolation === 'array') && isNumericRecord(from) && isNumericRecord(to)) {
    return interpolateNumericRecord(from, to, progress)
  }

  if (interpolation === 'color' && typeof from === 'string' && typeof to === 'string') {
    return interpolateColor(from, to, progress) ?? to
  }

  return to
}

function interpolateArray(from: readonly unknown[], to: readonly unknown[], progress: number): unknown[] {
  if (from.length !== to.length) {
    return [...to]
  }
  return from.map((value, index) => typeof value === 'number' && typeof to[index] === 'number'
    ? value + ((to[index] as number) - value) * progress
    : to[index])
}

function interpolateNumericRecord(from: unknown, to: unknown, progress: number): Record<string, unknown> {
  const left = from as Readonly<Record<string, unknown>>
  const right = to as Readonly<Record<string, unknown>>
  const result: Record<string, unknown> = { ...right }
  for (const key of Object.keys(right)) {
    const leftValue = left[key]
    const rightValue = right[key]
    result[key] = typeof leftValue === 'number' && typeof rightValue === 'number'
      ? leftValue + (rightValue - leftValue) * progress
      : rightValue
  }
  return result
}

function isNumericRecord(value: unknown): value is Readonly<Record<string, number>> {
  return Boolean(value && typeof value === 'object' && !Array.isArray(value))
}

export function easeProgress(progress: number, easing: string | undefined): number {
  const clamped = Math.max(0, Math.min(1, progress))
  switch (easing) {
    case 'linear':
      return clamped
    case 'ease':
      return cubicBezier(clamped, 0.25, 0.1, 0.25, 1)
    case 'ease-in':
    case 'easeIn':
      return clamped * clamped
    case 'ease-out':
    case 'easeOut':
      return 1 - (1 - clamped) * (1 - clamped)
    case 'ease-in-out':
    case 'easeInOut':
      return clamped < 0.5
        ? 2 * clamped * clamped
        : 1 - ((-2 * clamped + 2) ** 2) / 2
    case 'quad-in':
    case 'quadIn':
      return clamped * clamped
    case 'quad-out':
    case 'quadOut':
      return 1 - (1 - clamped) * (1 - clamped)
    case 'quad-in-out':
    case 'quadInOut':
      return clamped < 0.5 ? 2 * clamped * clamped : 1 - ((-2 * clamped + 2) ** 2) / 2
    case 'cubic-in':
    case 'cubicIn':
    case 'easeInCubic':
      return clamped ** 3
    case 'cubic-out':
    case 'cubicOut':
    case 'easeOutCubic':
      return 1 - (1 - clamped) ** 3
    case 'cubic-in-out':
    case 'cubicInOut':
    case 'easeInOutCubic':
      return clamped < 0.5 ? 4 * clamped ** 3 : 1 - ((-2 * clamped + 2) ** 3) / 2
    default:
      return parseCubicBezier(easing, clamped) ?? clamped
  }
}

function parseCubicBezier(easing: string | undefined, progress: number): number | undefined {
  const match = easing?.match(/^cubic-bezier\(([-\d.]+),\s*([-\d.]+),\s*([-\d.]+),\s*([-\d.]+)\)$/)
  if (!match)
    return undefined
  return cubicBezier(progress, Number(match[1]), Number(match[2]), Number(match[3]), Number(match[4]))
}

function cubicBezier(progress: number, _x1: number, y1: number, _x2: number, y2: number): number {
  const inverse = 1 - progress
  return 3 * inverse * inverse * progress * y1
    + 3 * inverse * progress * progress * y2
    + progress * progress * progress
}

function interpolateColor(from: string, to: string, progress: number): string | undefined {
  const left = parseColor(from)
  const right = parseColor(to)
  if (!left || !right)
    return undefined
  const rgba = left.map((value, index) => value + (right[index] - value) * progress)
  return rgba[3] >= 1
    ? `rgb(${Math.round(rgba[0])}, ${Math.round(rgba[1])}, ${Math.round(rgba[2])})`
    : `rgba(${Math.round(rgba[0])}, ${Math.round(rgba[1])}, ${Math.round(rgba[2])}, ${roundAlpha(rgba[3])})`
}

function parseColor(value: string): [number, number, number, number] | undefined {
  const hex = value.trim().match(/^#([\da-f]{3,8})$/i)
  if (hex) {
    return parseHexColor(hex[1])
  }
  const rgb = value.trim().match(/^rgba?\(([^)]+)\)$/i)
  if (!rgb)
    return undefined
  const parts = rgb[1].split(',').map(part => part.trim())
  if (parts.length < 3)
    return undefined
  return [
    Number(parts[0]),
    Number(parts[1]),
    Number(parts[2]),
    parts[3] === undefined ? 1 : Number(parts[3]),
  ]
}

function parseHexColor(hex: string): [number, number, number, number] | undefined {
  const normalized = hex.length === 3 || hex.length === 4
    ? hex.split('').map(char => `${char}${char}`).join('')
    : hex
  if (normalized.length !== 6 && normalized.length !== 8)
    return undefined
  return [
    Number.parseInt(normalized.slice(0, 2), 16),
    Number.parseInt(normalized.slice(2, 4), 16),
    Number.parseInt(normalized.slice(4, 6), 16),
    normalized.length === 8 ? Number.parseInt(normalized.slice(6, 8), 16) / 255 : 1,
  ]
}

function roundAlpha(value: number): number {
  return Math.round(value * 1000) / 1000
}

function resolveAt(at: AnimationTime, duration: number): number {
  if (typeof at === 'number')
    return at
  return Number.parseFloat(at) / 100 * duration
}

function setPath(target: Record<string, unknown>, property: string, value: unknown): void {
  const keys = property.split('.').filter(Boolean)
  if (keys.length === 0)
    return

  let cursor = target
  for (const key of keys.slice(0, -1)) {
    const current = cursor[key]
    if (!current || typeof current !== 'object' || Array.isArray(current)) {
      cursor[key] = {}
    }
    cursor = cursor[key] as Record<string, unknown>
  }
  cursor[keys[keys.length - 1]] = value
}

export function cloneUnknownRecord<T extends Readonly<Record<string, unknown>>>(value: T): Record<string, unknown> {
  return Object.fromEntries(Object.entries(value).map(([key, item]) => [key, cloneUnknownValue(item)]))
}

function cloneUnknownValue(value: unknown): unknown {
  if (Array.isArray(value)) {
    return value.map(cloneUnknownValue)
  }
  if (value && typeof value === 'object') {
    return cloneUnknownRecord(value as Readonly<Record<string, unknown>>)
  }
  return value
}

File: packages/native/crates/quajs_wgpu_renderer/src/projection_runtime/animation.rs
use serde_json::{Map, Value};

use super::{numeric_record, set_path};

/// Applies every resolved animation track to `view` and returns how many
/// animations are still running. The count feeds the dev performance HUD, and a
/// non-zero count means the renderer still owes more frames.
pub fn apply_animations(view: &mut Map<String, Value>, animations: &Value, now_ms: f64) -> usize {
    let Some(animations) = animations.as_array() else {
        return 0;
    };
    // Supplied timelines replace prior derived samples. Frames with no timeline
    // field may already have been sampled by the fixed-time TS serializer.
    if let Some(characters) = view.get_mut("characters").and_then(Value::as_array_mut) {
        for character in characters.iter_mut().filter_map(Value::as_object_mut) {
            character.remove("spriteLayerAnimationValues");
        }
    }
    let mut active = 0usize;
    for animation in animations {
        let Some(animation) = animation.as_object() else {
            continue;
        };
        if local_is_active(animation, now_ms) {
            active += 1;
        }
        let Some(elapsed) = local_time(animation, now_ms) else {
            continue;
        };
        let Some(tracks) = animation.get("resolvedTracks").and_then(Value::as_array) else {
            continue;
        };
        for track in tracks {
            let Some(track) = track.as_object() else {
                continue;
            };
            let Some(target) = track.get("target").and_then(Value::as_str) else {
                continue;
            };
            let Some(property) = track.get("property").and_then(Value::as_str) else {
                continue;
            };
            let Some(value) = track_value(
                track,
                elapsed,
                animation
                    .get("duration")
                    .and_then(Value::as_f64)
                    .unwrap_or(0.0),
            ) else {
                continue;
            };
            if target.starts_with("spriteLayer:") {
                crate::projection::character::animation::collect_sample(
                    view, target, property, &value,
                );
                continue;
            }
            for target_object in target_objects(view, target) {
                if target.starts_with("richText")
                    && !property.starts_with("style.")
                    && matches!(
                        property,
                        "color"
                            | "fontFamily"
                            | "fontSize"
                            | "fontWeight"
                            | "lineHeight"
                            | "textAlign"
                    )
                {
                    let value = if property == "lineHeight" && value.is_number() {
                        Value::String(value.to_string())
                    } else if property == "fontWeight" && value.is_number() {
                        // Native font selection uses integral OpenType weights.
                        Value::from(value.as_f64().unwrap().round().clamp(1.0, 1000.0) as u16)
                    } else if property == "fontFamily" && value.is_string() {
                        Value::Array(vec![value.clone()])
                    } else {
                        value.clone()
                    };
                    set_path(target_object, &format!("style.{property}"), value);
                } else {
                    set_path(target_object, property, value.clone());
                }
            }
        }
    }
    active
}

fn target_objects<'a>(
    view: &'a mut Map<String, Value>,
    target: &str,
) -> Vec<&'a mut Map<String, Value>> {
    // Match the shared projection roots. Rich text may select the same span id
    // in multiple blocks and must update visible and full-layout documents.
    let mut result = Vec::new();
    if target.starts_with("richText") {
        let parts: Vec<_> = target.split(':').collect();
        let fields = match parts.get(1).copied() {
            Some("dialogue") => ["text", "layoutText"],
            Some("speaker") => ["speaker", "layoutSpeaker"],
            _ => return result,
        };
        let Some(dialogue) = view.get_mut("dialogue").and_then(Value::as_object_mut) else {
            return result;
        };
        for (key, value) in dialogue.iter_mut() {
            if fields.contains(&key.as_str()) {
                result.extend(rich_text_targets(value, &parts));
            }
        }
    } else if matches!(target, "stage:main" | "camera:main") {
        let key = target.split(':').next().unwrap();
        let root = view.entry(key).or_insert_with(|| Value::Object(Map::new()));
        if let Some(object) = root.as_object_mut() {
            result.push(object);
        }
    } else if target == "background:main" {
        if let Some(object) = view.get_mut("background").and_then(Value::as_object_mut) {
            result.push(object);
        }
    } else if let Some(id) = target.strip_prefix("backgroundLayer:") {
        if let Some(layers) = view
            .get_mut("background")
            .and_then(|value| value.get_mut("layers"))
            .and_then(Value::as_array_mut)
        {
            if let Some(object) = layers
                .iter_mut()
                .find(|layer| layer.get("id").and_then(Value::as_str) == Some(id))
                .and_then(Value::as_object_mut)
            {
                result.push(object);
            }
        }
    } else if let Some(id) = target.strip_prefix("character:") {
        if let Some(characters) = view.get_mut("characters").and_then(Value::as_array_mut) {
            if let Some(object) = characters
                .iter_mut()
                .find(|character| character.get("id").and_then(Value::as_str) == Some(id))
                .and_then(Value::as_object_mut)
            {
                result.push(object);
            }
        }
    } else if target == "dialogue:box" {
        if let Some(object) = view.get_mut("dialogue").and_then(Value::as_object_mut) {
            result.push(object);
        }
    } else if target == "choices:panel" {
        if let Some(object) = view.get_mut("choices").and_then(Value::as_object_mut) {
            result.push(object);
        }
    } else if let Some(id) = target.strip_prefix("choice:") {
        if let Some(choices) = view
            .get_mut("choices")
            .and_then(|value| value.get_mut("choices"))
            .and_then(Value::as_array_mut)
        {
            if let Some(object) = choices
                .iter_mut()
                .find(|choice| choice.get("id").and_then(Value::as_str) == Some(id))
                .and_then(Value::as_object_mut)
            {
                result.push(object);
            }
        }
    } else if let Some(id) = target.strip_prefix("effect:") {
        if let Some(effects) = view.get_mut("effects").and_then(Value::as_array_mut) {
            if let Some(object) = effects
                .iter_mut()
                .find(|effect| effect.get("id").and_then(Value::as_str) == Some(id))
                .and_then(Value::as_object_mut)
            {
                result.push(object);
            }
        }
    } else if let Some(id) = target.strip_prefix("ui:") {
        if let Some(overlays) = view
            .get_mut("ui")
            .and_then(|value| value.get_mut("overlays"))
            .and_then(Value::as_array_mut)
        {
            if let Some(object) = overlays
                .iter_mut()
                .find(|overlay| overlay.get("elementId").and_then(Value::as_str) == Some(id))
                .and_then(Value::as_object_mut)
            {
                result.push(object);
            }
        }
    }
    result
}

fn rich_text_targets<'a>(value: &'a mut Value, parts: &[&str]) -> Vec<&'a mut Map<String, Value>> {
    let mut result = Vec::new();
    let Some(document) = value.as_object_mut() else {
        return result;
    };
    if parts == ["richText", "dialogue"] || parts == ["richText", "speaker"] {
        result.push(document);
        return result;
    }
    let Some(id) = parts
        .get(2..)
        .filter(|p| !p.is_empty())
        .map(|p| p.join(":"))
    else {
        return result;
    };
    let Some(blocks) = document.get_mut("blocks").and_then(Value::as_array_mut) else {
        return result;
    };
    for (index, block) in blocks.iter_mut().enumerate() {
        let Some(block) = block.as_object_mut() else {
            continue;
        };
        if parts[0] == "richTextBlock" {
            if rich_text_id_matches(block, index, &id) {
                result.push(block);
            }
        } else if parts[0] == "richTextSpan" {
            // Shared render-core targets a span id (or its index within each
            // block), without a block-id segment. Repeated ids match each block.
            if let Some(spans) = block.get_mut("spans").and_then(Value::as_array_mut) {
                for (index, span) in spans.iter_mut().enumerate() {
                    if let Some(span) = span
                        .as_object_mut()
                        .filter(|s| rich_text_id_matches(s, index, &id))
                    {
                        result.push(span);
                    }
                }
            }
        }
    }
    result
}

fn rich_text_id_matches(object: &Map<String, Value>, index: usize, id: &str) -> bool {
    object
        .get("id")
        .and_then(Value::as_str)
        .filter(|id| !id.is_empty())
        .map_or_else(|| index.to_string() == id, |value| value == id)
}

fn local_time(animation: &Map<String, Value>, now_ms: f64) -> Option<f64> {
    let started = animation.get("startedAt").and_then(Value::as_f64)?;
    let duration = animation
        .get("duration")
        .and_then(Value::as_f64)
        .unwrap_or(0.0)
        .max(0.0);
    let playback_rate = animation
        .get("playbackRate")
        .and_then(Value::as_f64)
        .filter(|value| value.is_finite() && *value > 0.0)
        .unwrap_or(1.0);
    let delay = animation
        .get("delay")
        .and_then(Value::as_f64)
        .unwrap_or(0.0)
        .max(0.0);
    if animation.get("state").and_then(Value::as_str) == Some("stopped")
        && animation.get("endedAt").and_then(Value::as_f64).is_none()
    {
        return None;
    }
    let wall = if animation.get("state").and_then(Value::as_str) == Some("paused") {
        animation
            .get("pausedAt")
            .and_then(Value::as_f64)
            .unwrap_or(now_ms)
            - started
    } else if animation.get("state").and_then(Value::as_str) == Some("stopped") {
        animation
            .get("endedAt")
            .and_then(Value::as_f64)
            .unwrap_or(now_ms)
            - started
    } else {
        now_ms - started
    };
    let timeline = (wall * playback_rate).max(0.0);
    if timeline < delay {
        return if matches!(
            animation.get("fill").and_then(Value::as_str),
            Some("backwards") | Some("both")
        ) {
            Some(direction(0.0, duration, 0, animation))
        } else {
            None
        };
    }
    if duration <= 0.0 {
        return Some(0.0);
    }
    let active = timeline - delay;
    let loops = match animation.get("loop") {
        Some(Value::Bool(true)) => f64::INFINITY,
        Some(Value::Number(value)) => value.as_f64().unwrap_or(1.0).max(1.0),
        _ => 1.0,
    };
    if active >= duration * loops {
        return if matches!(
            animation
                .get("fill")
                .and_then(Value::as_str)
                .unwrap_or("forwards"),
            "forwards" | "both"
        ) {
            Some(direction(duration, duration, loops as usize - 1, animation))
        } else {
            None
        };
    }
    let iteration = (active / duration).floor() as usize;
    Some(direction(
        active - iteration as f64 * duration,
        duration,
        iteration,
        animation,
    ))
}

fn local_is_active(animation: &Map<String, Value>, now_ms: f64) -> bool {
    let Some(started) = animation.get("startedAt").and_then(Value::as_f64) else {
        return false;
    };
    if matches!(
        animation.get("state").and_then(Value::as_str),
        Some("paused") | Some("stopped")
    ) {
        return false;
    }
    let duration = animation
        .get("duration")
        .and_then(Value::as_f64)
        .unwrap_or(0.0)
        .max(0.0);
    let delay = animation
        .get("delay")
        .and_then(Value::as_f64)
        .unwrap_or(0.0)
        .max(0.0);
    let loops = match animation.get("loop") {
        Some(Value::Bool(true)) => f64::INFINITY,
        Some(Value::Number(value)) => value.as_f64().unwrap_or(1.0).max(1.0),
        _ => 1.0,
    };
    let playback_rate = animation
        .get("playbackRate")
        .and_then(Value::as_f64)
        .filter(|value| value.is_finite() && *value > 0.0)
        .unwrap_or(1.0);
    now_ms < started + (delay + duration * loops) / playback_rate
}

fn direction(local: f64, duration: f64, iteration: usize, animation: &Map<String, Value>) -> f64 {
    let direction = animation
        .get("direction")
        .and_then(Value::as_str)
        .unwrap_or("normal");
    let reversed = direction == "reverse"
        || (direction == "alternate" && iteration % 2 == 1)
        || (direction == "alternate-reverse" && iteration % 2 == 0);
    if reversed {
        duration - local
    } else {
        local
    }
}

fn track_value(track: &Map<String, Value>, elapsed: f64, duration: f64) -> Option<Value> {
    let frames = track.get("keyframes").and_then(Value::as_array)?;
    if frames.is_empty() {
        return None;
    }
    if frames.len() == 1 {
        return frames[0].get("value").cloned();
    }
    let mut frames = frames.clone();
    frames.sort_by(|left, right| {
        at(left.get("at"), duration)
            .partial_cmp(&at(right.get("at"), duration))
            .unwrap_or(std::cmp::Ordering::Equal)
    });
    if elapsed <= at(frames[0].get("at"), duration) {
        return frames[0].get("value").cloned();
    }
    for index in 1..frames.len() {
        let previous = &frames[index - 1];
        let next = &frames[index];
        let previous_at = at(previous.get("at"), duration);
        let next_at = at(next.get("at"), duration);
        if elapsed <= next_at {
            if elapsed == next_at {
                return next.get("value").cloned();
            }
            if matches!(
                track.get("interpolation").and_then(Value::as_str),
                Some("step") | Some("discrete")
            ) {
                return previous.get("value").cloned();
            }
            let progress = if next_at == previous_at {
                1.0
            } else {
                (elapsed - previous_at) / (next_at - previous_at)
            };
            return interpolate(
                previous.get("value")?,
                next.get("value")?,
                timeline_ease_progress(
                    progress,
                    next.get("easing")
                        .and_then(Value::as_str)
                        .or_else(|| previous.get("easing").and_then(Value::as_str)),
                ),
                track.get("interpolation").and_then(Value::as_str),
            );
        }
    }
    frames.last().and_then(|frame| frame.get("value").cloned())
}

fn at(value: Option<&Value>, duration: f64) -> f64 {
    match value {
        Some(Value::Number(number)) => number.as_f64().unwrap_or(0.0),
        Some(Value::String(value)) => {
            value.trim_end_matches('%').parse::<f64>().unwrap_or(0.0) / 100.0 * duration
        }
        _ => 0.0,
    }
}

fn interpolate(
    from: &Value,
    to: &Value,
    progress: f64,
    interpolation: Option<&str>,
) -> Option<Value> {
    let p = progress.clamp(0.0, 1.0);
    if let (Some(left), Some(right)) = (from.as_f64(), to.as_f64()) {
        return Some(Value::from(left + (right - left) * p));
    }
    if matches!(interpolation, Some("array") | Some("vector")) {
        if let (Some(left), Some(right)) = (from.as_array(), to.as_array()) {
            if left.len() != right.len() {
                return Some(to.clone());
            }
            return Some(Value::Array(
                left.iter()
                    .zip(right)
                    .map(|(a, b)| interpolate(a, b, p, None).unwrap_or_else(|| b.clone()))
                    .collect(),
            ));
        }
        if numeric_record(from) && numeric_record(to) {
            let mut result = to.as_object().cloned().unwrap_or_default();
            for (key, right) in to.as_object()? {
                if let (Some(left), Some(right)) =
                    (from.get(key).and_then(Value::as_f64), right.as_f64())
                {
                    result.insert(key.clone(), Value::from(left + (right - left) * p));
                }
            }
            return Some(Value::Object(result));
        }
    }
    if interpolation == Some("color") {
        if let (Some(left), Some(right)) = (parse_color(from.as_str()?), parse_color(to.as_str()?))
        {
            let rgba = left
                .iter()
                .zip(right)
                .map(|(left, right)| left + (right - left) * p)
                .collect::<Vec<_>>();
            let alpha = rgba[3];
            if alpha >= 0.999 {
                return Some(Value::String(format!(
                    "rgb({}, {}, {})",
                    rgba[0].round() as i64,
                    rgba[1].round() as i64,
                    rgba[2].round() as i64
                )));
            }
            return Some(Value::String(format!(
                "rgba({}, {}, {}, {:.3})",
                rgba[0].round() as i64,
                rgba[1].round() as i64,
                rgba[2].round() as i64,
                alpha
            )));
        }
    }
    Some(to.clone())
}

fn parse_color(value: &str) -> Option<[f64; 4]> {
    let value = value.trim();
    if let Some(hex) = value.strip_prefix('#') {
        let expanded = match hex.len() {
            3 | 4 => hex
                .chars()
                .flat_map(|char| [char, char])
                .collect::<String>(),
            6 | 8 => hex.to_string(),
            _ => return None,
        };
        let channel = |start: usize| {
            u8::from_str_radix(&expanded[start..start + 2], 16)
                .ok()
                .map(f64::from)
        };
        return Some([
            channel(0)?,
            channel(2)?,
            channel(4)?,
            if expanded.len() == 8 {
                channel(6)? / 255.0
            } else {
                1.0
            },
        ]);
    }
    let body = value
        .strip_prefix("rgb(")
        .or_else(|| value.strip_prefix("rgba("))
        .and_then(|value| value.strip_suffix(')'))?;
    let channels = body.split(',').map(str::trim).collect::<Vec<_>>();
    if channels.len() < 3 {
        return None;
    }
    Some([
        channels[0].parse().ok()?,
        channels[1].parse().ok()?,
        channels[2].parse().ok()?,
        channels
            .get(3)
            .and_then(|value| value.parse().ok())
            .unwrap_or(1.0),
    ])
}

// Engine timeline keywords follow render-core. CSS interaction/presence easing
// still uses the CSS solver below; the two contracts deliberately differ.
fn timeline_ease_progress(progress: f64, easing: Option<&str>) -> f64 {
    ease_progress(
        progress,
        match easing {
            Some("ease-in") => Some("quad-in"),
            Some("ease-out") => Some("quad-out"),
            Some("ease-in-out") => Some("quad-in-out"),
            _ => easing,
        },
    )
}

pub fn ease_progress(progress: f64, easing: Option<&str>) -> f64 {
    let p = progress.clamp(0.0, 1.0);
    let Some(easing) = easing else {
        return p;
    };
    // CSS standard keywords and cubic-bezier / steps functions delegate to the
    // shared CSS-spec solver so that both the timeline path and the
    // interaction-feedback transition path produce identical bezier outputs.
    if let Some(value) = super::easing::ease_for_css_keyword(p, easing) {
        return value;
    }
    // render-core / legacy aliases that have no CSS standard equivalent.
    match easing {
        "easeIn" | "quad-in" | "quadIn" => p * p,
        "easeOut" | "quad-out" | "quadOut" => 1.0 - (1.0 - p) * (1.0 - p),
        "easeInOut" | "quad-in-out" | "quadInOut" => {
            if p < 0.5 {
                2.0 * p * p
            } else {
                1.0 - ((-2.0 * p + 2.0).powi(2)) / 2.0
            }
        }
        "cubic-in" | "cubicIn" | "easeInCubic" => p.powi(3),
        "cubic-out" | "cubicOut" | "easeOutCubic" => 1.0 - (1.0 - p).powi(3),
        "cubic-in-out" | "cubicInOut" | "easeInOutCubic" => {
            if p < 0.5 {
                4.0 * p.powi(3)
            } else {
                1.0 - ((-2.0 * p + 2.0).powi(3)) / 2.0
            }
        }
        // Unknown keyword → linear, matching render-core fallback.
        _ => p,
    }
}
